fdb-record-layer-lucene/src/main/java/com/apple/foundationdb/record/lucene/codec/LuceneOptimizedFieldInfosFormat.java

```diff
@@ -20,36 +20,277 @@
 
 package com.apple.foundationdb.record.lucene.codec;
 
+import com.apple.foundationdb.record.RecordCoreException;
+import com.apple.foundationdb.record.lucene.LuceneFieldInfosProto;
+import com.apple.foundationdb.record.lucene.LuceneLogMessageKeys;
+import com.apple.foundationdb.record.lucene.directory.FDBDirectory;
+import com.apple.foundationdb.record.lucene.directory.FDBLuceneFileReference;
+import com.google.protobuf.ByteString;
+import com.google.protobuf.InvalidProtocolBufferException;
 import org.apache.lucene.codecs.FieldInfosFormat;
+import org.apache.lucene.index.DocValuesType;
+import org.apache.lucene.index.FieldInfo;
 import org.apache.lucene.index.FieldInfos;
+import org.apache.lucene.index.IndexFileNames;
+import org.apache.lucene.index.IndexOptions;
 import org.apache.lucene.index.SegmentInfo;
 import org.apache.lucene.store.Directory;
+import org.apache.lucene.store.FilterDirectory;
 import org.apache.lucene.store.IOContext;
 
+import javax.annotation.Nonnull;
+import java.io.FileNotFoundException;
 import java.io.IOException;
+import java.util.ArrayList;
+import java.util.BitSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.function.Function;
+import java.util.stream.Collectors;
 
 /**
- * This class attempts to reduce the number of .fnm entries when numerous segments are present in FDB.
+ * {@link FieldInfosFormat} optimized for storage in the {@link FDBDirectory}.
+ * <p>
+ *     The key feature here is that it will store on the file reference a reference to a (potentially) shared protobuf
+ *     for the FieldInfos, along with a bitset for the subset of the fields in that shared proto that are used in the
+ *     associated segment. This deduplication is important because segments generally have the same mapping (there is a
+ *     global mapping used when creating new segments), but we also need to support different segments having
+ *     incompatible mappings. See <a href="https://github.com/FoundationDB/fdb-record-layer/issues/2284">Issue #2284</a>
+ *     for more information about why we need to support incompatible mappings.
+ * </p>
  */
 public class LuceneOptimizedFieldInfosFormat extends FieldInfosFormat {
-    private final FieldInfosFormat fieldInfosFormat;
 
-    public LuceneOptimizedFieldInfosFormat(FieldInfosFormat fieldInfosFormat) {
-        this.fieldInfosFormat = fieldInfosFormat;
-    }
+    public static final String EXTENSION = "fip";
```

Comment:
My intent was purely for maintainability purposes, so looking at a file you can tell that `fdb-si` is not a standard Lucene file extension. Having an enum is a good way to capture that since there would be a way to see them all in one place.